apv5sdk-v15/apps/int6000-utils-1.1.6/pib/getpib.7
.TH getpib 7 "Intellon Corporation" "int6000-utils-linux" "Intellon Linux Toolkit"
.SH NAME
getpib - Intellon PIB Data Extractor 
.SH SYNOPSIS
.BR getpib 
.IR file 
.IR offset 
.IR format 
.RI [ length ]
.SH DESCRIPTION
This program is a companion to program \fBsetpib\fR and is designed to support shell scripts. It extracts PIB file parameters so that a scripting language can evaluate, manipulate or store them with minimal effort. Good scripting skills and some imagination may be required to make effective use of this program.
.PP
Users having a detailed knowledge of PIB content and structure can make extremely effective use of this program; however, Intellon is under no obligation to provide customers with the offset, length or format of PIB file parameters.
.PP
This program is part of the Intellon Linux Toolkit. See the \fBilt\fR man page for an overview and installation instructions.
.SH OPTIONS
.TP
.RB - q
Enable quiet mode which has no effect.
.TP
.RB - v
Print a dump of the edited region. The extent of displayed data can be changed using option \fB-w\fR.
.TP
.RB - ? ,-- help
Print program help summary on stdout. This option takes precedence over other options on the command line. 
.TP
.RB - ! ,-- version
Print program version information on stdout. This option takes precedence over other options on the command line. Usethis option when sending screen dumps to Intellon Technical Support so that they know exactly which version of the Linux Toolkit you are using.
.SH ARGUMENTS
.TP
.IR file
The PIB filename. This argument is required and must appear first on the command line. No assumptions are made based on filename and no filename conventions are enforced; however, an invalid PIB file will be rejected.
.TP
.IR offset
The data offset in bytes expressed in hexadecimal. This argument is required and must follow the filename. A leading "\fB0x\fR" prefix is optional. The data \fIoffset\fR plus the data \fIlength\fR cannot exceed the file extent.
.TP
.IB format
The output format. This argument is required and must follow the offset. Some formats imply fixed data length. Others formats are for variable length data. The fixed data formats are \fBbyte\fR, \fBword\fR, \fBlong\fR, \fBhfid\fR, \fBmac\fR and \fBkey\fR. The variable data formats are \fBdata\fR and \fBtext\fR. The formats are described in the next section.
.TP
.IB length
The data length in bytes expressed in decimal. This argument is only required for the the data formats \fBdata\fR and \fBtext\fR. The data \fIoffset\fR plus the data \fIlength\fR cannot exceed the file extent. The data \fIlength\fR cannot exceed \fB1024\fR.
.SH FORMATS
.TP
.BR byte
Extract 1 byte and display it as an unsigned decimal integer string. The offset advances \f1\fR byte before starting another extraction. 
.TP
.BR word
Extract 2 consecutive bytes and display them as an unsigned decimal integer string after endian conversion.  The offset advances \f2\fR bytes before starting another extraction. 
.TP
.BR long
Extract 4 consecutive bytes and display them as an unsigned decimal integer string after endian conversion.  The offset advances \fB4\fR bytes before starting another extraction. 
.TP
.BR hfid
Extract 64 consecutive bytes and display them as an ASCII string that terminates at the first non-printable character. The offset advances \fB64\fR bytes before starting another extraction. This format is equivalent to "\fBtext 64\fR".  
.TP
.BR mac
Extract 6 bytes and display them as a colon-separated hexadecimal string. The offset advances \f6\fR bytes before starting another extraction. This is similar to "\fBdata 6\fR" but colons separate octets. 
.TP
.BR key
Extract 16 bytes and display them as a colon-separated hexadecimal string. The offset advances \fI16\fR bytes before starting another extraction. This is similar to "\fBdata 16\fR" but colons separate octets. 
.TP
\fBdata \fIlength\fR
Extract the specified number of bytes and display them as a hexadecimal string with no punctuation to separate octets. The offset advances \fIlength\fR bytes before starting another extraction. The minimum length is \fB1\fR. The maximum length is \fB1024\fR.
.TP
\fBtext \fIlength\fR
Extract the specified number of bytes and display them as an ASCII string terminated at first non-printable character. The offset advances \fIlength\fR bytes before starting another extraction. The minimum length is \fB1\fR. The maximum length is \fB1024\fR.
.TP
\fBskip \fIlength\fR
Advance the specified number of bytes without displaying anything. The offset advances by \fIlength\fR bytes before starting another extraction. The minimum length is \fB1\fR. The maximum length is \fB1024\fR.
.SH EXAMPLES
The following example extracts one byte from offset \fB01F7\fR of PIB file \fBabc.pib\fR and displays it as an unsigned decimal integer string. No length specification is needed because the \fBbyte\fR format has an implied length of \fB1\fR byte. The displayed value is \fB232\fR because the \fBbyte\fR format is decimal. We could have specified "\fBdata 1\fR" to display the byte in hexadecimal. The return prompt appears on the display line because option \fB-n\fR was omitted. 
.PP
   # getpib abc.pib 01F7 byte 
   232#
.PP
The next example extracts two bytes at offset \fB01F7\fR and displays them as a hexadecimal string. A length of \fB2\fR is needed because the \fBdata\fR format is variable length. We could have specified "\fBword\fR" to display these bytes as an unsigned decimal integer string. The hexadecimal string consists two octets \fBE8\fR and \fB8A\fR. The first byte is the same one extracted in the last example. The return prompt appears on a new line because option \fB-n\fR was present.
.PP
   # getpib abc.pib 01F7 data 2 -n
   E88A
   #
.SH DISCLAIMER
PIB file structure and content is proprietary to Intellon Corporation, Ocala FL USA. Consequently, public information is not available. Intellon Corporation reserves the right to modify PIB file structure or content in future firmware releases without any obligation to notify or compensate users of this program.
.SH SEE ALSO
.BR chkpib ( 7 ),
.BR modpib ( 7 ),
.BR mrgpib ( 7 ),
.BR setpib ( 7 ),
.BR pxe ( 7 )
.SH CREDITS
 Charles Maier <charles.maier@intellon.com>


apv5sdk-v15/autelan/plc-homeplug/pib/getpib.c
/*====================================================================*
 *   
 *   Copyright (c) 2011 by Qualcomm Atheros.
 *   
 *   Permission to use, copy, modify, and/or distribute this software 
 *   for any purpose with or without fee is hereby granted, provided 
 *   that the above copyright notice and this permission notice appear 
 *   in all copies.
 *   
 *   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL 
 *   WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED 
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL  
 *   THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR 
 *   CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM 
 *   LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, 
 *   NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN 
 *   CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *   
 *--------------------------------------------------------------------*/

/*====================================================================*
 *
 *   getpib.c - PIB Data Extractor
 *
 *   This software and documentation is the property of Atheros 
 *   Corporation, Ocala, Florida. It is provided 'as is' without 
 *   expressed or implied warranty of any kind to anyone for any 
 *   reason. Atheros assumes no responsibility or liability for 
 *   errors or omissions in the software or documentation and 
 *   reserves the right to make changes without notification. 
 *   
 *   Atheros customers may modify and distribute the software 
 *   without obligation to Atheros. Since use of this software 
 *   is optional, users shall bear sole responsibility and 
 *   liability for any consequences of it's use. 
 *   
 *.  Qualcomm Atheros HomePlug AV Powerline Toolkit
 *:  Published 2009-2011 by Qualcomm Atheros. ALL RIGHTS RESERVED
 *;  For demonstration and evaluation only. Not for production use
 *
 *   Contributor(s):
 *      Charles Maier <cmaier@qualcomm.com>
 *
 *--------------------------------------------------------------------*/

#define _GETOPT_H

/*====================================================================*"
 *   system header files;
 *--------------------------------------------------------------------*/

#include <unistd.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>

/*====================================================================*
 *   custom header files;
 *--------------------------------------------------------------------*/

#include "../tools/getoptv.h"
#include "../tools/memory.h"
#include "../tools/number.h"
#include "../tools/error.h"
#include "../tools/types.h"
#include "../tools/flags.h"
#include "../tools/files.h"
#include "../pib/pib.h"

/*====================================================================*
 *   custom source files;
 *--------------------------------------------------------------------*/

#ifndef MAKEFILE
#include "../tools/getoptv.c"
#include "../tools/putoptv.c"
#include "../tools/version.c"
#include "../tools/uintspec.c"
#include "../tools/basespec.c"
#include "../tools/todigit.c"
#include "../tools/hexout.c"
#include "../tools/error.c"
#endif

/*====================================================================*
 *   constants;
 *--------------------------------------------------------------------*/

#define SETPIB_VERBOSE (1 << 0)
#define SETPIB_SILENCE (1 << 1)
#define SETPIB_NEWLINE (1 << 2)

/*====================================================================*
 *
 *   void getmemory (byte const * memory, size_t extent, char const * object, size_t length);
 *
 *
 *--------------------------------------------------------------------*/

static void getmemory (byte const * memory, size_t extent, char const * object, size_t length) 

{
	if (length > extent) 
	{
		error (1, ECANCELED, "%s exceeds PIB length of " SIZE_T_SPEC " bytes", object, length);
	}
	hexout (memory, length, ':', 0, stdout);
	return;
}


/*====================================================================*
 *
 *   void getstring (byte const * memory, size_t extent, char const * object, size_t length);
 *
 *
 *--------------------------------------------------------------------*/

static void getstring (byte const * memory, size_t extent, char const * object, size_t length) 

{
	char const * string = (char const *)(memory);
	if (length > extent) 
	{
		error (1, ECANCELED, "%s exceeds PIB length " SIZE_T_SPEC " bytes", object, length);
	}
	while (isprint (*string) && (length--)) 
	{
		putc (*string++, stdout);
	}
	return;
}


/*====================================================================*
 *
 *   int main (int argc, char const * argv []);
 *   
 *.  Qualcomm Atheros HomePlug AV Powerline Toolkit
 *:  Published 2009-2011 by Qualcomm Atheros. ALL RIGHTS RESERVED
 *;  For demonstration and evaluation only. Not for production use
 *
 *--------------------------------------------------------------------*/

int main (int argc, char const * argv []) 

{
	static char const * optv [] = 
	{
		"qvn",
		"file offset type [size]\n\n\t  standard-length types are 'byte'|'number'|'long'|'hfid'|'mac'|'key'\n\t  variable-length types are 'data'|'text'",
		"PIB Data Extractor",
		"n\tappend newline",
		"q\tquiet mode",
		"v\tverbose mode",
		(char const *) (0)
	};
	struct _file_ file;
	struct pib_header * header;
	unsigned offset = 0;
	unsigned length = 0;
	unsigned extent = 0;
	byte * buffer;
	byte const * memory;
	flag_t flags = (flag_t)(0);
	signed c;
	optind = 1;
	opterr = 1;
	while ((c = getoptv (argc, argv, optv)) != -1) 
	{
		switch (c) 
		{
		case 'n':
			_setbits (flags, SETPIB_NEWLINE);
			break;
		case 'q':
			_setbits (flags, SETPIB_SILENCE);
			break;
		case 'v':
			_setbits (flags, SETPIB_VERBOSE);
			break;
		default:
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (!argc) 
	{
		error (1, 0, "No file to read");
	}
	file.name = * argv;
	if ((file.file = open (file.name, O_BINARY|O_RDONLY)) == -1) 
	{
		error (1, errno, FILE_CANTOPEN, file.name);
	}
	if ((extent = lseek (file.file, 0, SEEK_END)) == (unsigned)(-1)) 
	{
		error (1, errno, FILE_CANTSIZE, file.name);
	}
	if (!(buffer = malloc (extent))) 
	{
		error (1, errno, "Can't span %s", file.name);
	}
	if (lseek (file.file, 0, SEEK_SET)) 
	{
		error (1, errno, FILE_CANTHOME, file.name);
	}
	if (read (file.file, buffer, extent) != (signed)(extent)) 
	{
		error (1, errno, FILE_CANTLOAD, file.name);
	}
	close (file.file);
	argc--;
	argv++;
	header = (struct pib_header *)(buffer);
	if (extent != LE16TOH (header->PIBLENGTH)) 
	{
		error (1, ECANCELED, "Bad PIB filesize: %s", file.name);
	}
	if (!argc) 
	{
		error (1, ECANCELED, "Need an offset");
	}
	length = (uint32_t)(basespec (* argv, 16, sizeof (uint32_t)));
	if (length > extent) 
	{
		error (1, ECANCELED, "PIB offset %X exceeds PIB extent %X", offset, length);
	}
	memory = buffer + length;
	argc--;
	argv++;
	if (!argc) 
	{
		_setbits (flags, SETPIB_VERBOSE);
	}
	while (argc && * argv) 
	{
		char const * object = * argv;
		argc--;
		argv++;
		if (!strcmp (object, "byte")) 
		{
			uint8_t * number = (uint8_t *)(memory);
			if (sizeof (* number) > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			printf ("%u", * number);
			memory += sizeof (* number);
			extent -= sizeof (* number);
		}
		else if (!strcmp (object, "word")) 
		{
			uint16_t * number = (uint16_t *)(memory);
			if (sizeof (* number) > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			printf ("%u", LE16TOH (* number));
			memory += sizeof (* number);
			extent -= sizeof (* number);
		}
		else if (!strcmp (object, "long")) 
		{
			uint32_t * number = (uint32_t *)(memory);
			if (sizeof (* number) > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			printf ("%u", LE32TOH (* number));
			memory += sizeof (* number);
			extent -= sizeof (* number);
		}

#if 1

		else if (!strcmp (object, "xbyte")) 
		{
			uint8_t * number = (uint8_t *)(memory);
			if (sizeof (* number) > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			printf ("0x%02X", * number);
			memory += sizeof (* number);
			extent -= sizeof (* number);
		}
		else if (!strcmp (object, "xword")) 
		{
			uint16_t * number = (uint16_t *)(memory);
			if (sizeof (* number) > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			printf ("0x%04X", LE16TOH (* number));
			memory += sizeof (* number);
			extent -= sizeof (* number);
		}
		else if (!strcmp (object, "xlong")) 
		{
			uint32_t * number = (uint32_t *)(memory);
			if (sizeof (* number) > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			printf ("0x%08X", LE32TOH (* number));
			memory += sizeof (* number);
			extent -= sizeof (* number);
		}

#endif

		else if (!strcmp (object, "mac")) 
		{
			length = ETHER_ADDR_LEN;
			if (length > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			getmemory (memory, extent, object, length);
			memory += length;
			extent -= length;
		}
		else if (!strcmp (object, "key")) 
		{
			length = PIB_KEY_LEN;
			if (length > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			getmemory (memory, extent, object, length);
			memory += length;
			extent -= length;
		}
		else if (!strcmp (object, "hfid")) 
		{
			length = PIB_HFID_LEN;
			if (length > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			getstring (memory, extent, object, length);
			memory += length;
			extent -= length;
		}

#if 1

		else if (!strcmp (object, "adminusername") || !strcmp (object, "adminpassword") || !strcmp (object, "accessusername")) 
		{
			length = PIB_NAME_LEN + 1;
			if (length > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			getstring (memory, extent, object, length);
			memory += length;
			extent -= length;
		}
		else if (!strcmp (object, "accesspassword")) 
		{
			length = PIB_HFID_LEN + 1;
			if (length > extent) 
			{
				error (1, ECANCELED, "%s exceeds PIB extent %X", object, extent);
			}
			getstring (memory, extent, object, length);
			memory += length;
			extent -= length;
		}
		else if (!strcmp (object, "username") || !strcmp (object, "password") || !strcmp (object, "url")) 
		{
			length = PIB_TEXT_LEN + 1;
			getstring (memory, extent, object, length);
			memory += length;
			extent -= length;
		}

#endif

		else if (!strcmp (object, "data")) 
		{
			if (!* argv) 
			{
				error (1, EINVAL, "%s needs a length", object);
			}
			length = (unsigned)(uintspec (* argv, 1, extent));
			hexout (memory, length, 0, 0, stdout);
			memory += length;
			extent -= length;
			argc--;
			argv++;
		}
		else if (!strcmp (object, "text")) 
		{
			if (!* argv) 
			{
				error (1, EINVAL, "%s needs a length", object);
			}
			length = (unsigned)(uintspec (* argv, 1, extent));
			getstring (memory, extent, object, length);
			memory += length;
			extent -= length;
			argc--;
			argv++;
		}
		else if (!strcmp (object, "skip")) 
		{
			if (!* argv) 
			{
				error (1, EINVAL, "%s needs a length", object);
			}
			length = (unsigned)(uintspec (* argv, 1, extent));
			memory += length;
			extent -= length;
			argc--;
			argv++;
			continue;
		}
		else 
		{
			error (1, ENOTSUP, "%s", object);
		}
		if ((argc) && (* argv)) 
		{
			putc (' ', stdout);
		}
	}
	if (_anyset (flags, SETPIB_NEWLINE)) 
	{
		putc ('\n', stdout);
	}
	free (buffer);
	return (0);
}

